src/OrchardCore.Modules/OrchardCore.Media/Processing/MediaTokenService.cs

```diff
@@ -1,44 +1,45 @@
 using System;
-using System.Collections.Concurrent;
 using System.Collections.Generic;
-using System.Linq;
-using System.Web;
-using Microsoft.AspNetCore.DataProtection;
+using System.Security.Cryptography;
+using System.Text;
 using Microsoft.AspNetCore.WebUtilities;
 using Microsoft.Extensions.Caching.Memory;
-using Microsoft.Extensions.Primitives;
-using Newtonsoft.Json;
-using OrchardCore.Modules;
+using OrchardCore.Entities;
+using OrchardCore.Settings;
 using SixLabors.ImageSharp.Web.Processors;
 
 namespace OrchardCore.Media.Processing
 {
     public class MediaTokenService : IMediaTokenService
     {
-        private readonly IDataProtector _dataProtector;
+        private const string TokenCacheKeyPrefix = "MediaToken:";
         private readonly IMemoryCache _memoryCache;
-        private readonly HashSet<string> _knownCommands = new HashSet<string>();
+        private readonly ISiteService _siteService;
 
-        private const string TokenCacheKeyPrefix = "MediaToken:";
-        private const string CommandCacheKeyPrefix = "MediaCommands:";
+        private readonly HashSet<string> _knownCommands = new HashSet<string>();
+        private readonly byte[] _hashKey;
 
         public MediaTokenService(
-            IDataProtectionProvider dataProtectionProvider,
             IMemoryCache memoryCache,
+            ISiteService siteService,
             IEnumerable<IImageWebProcessor> processors)
         {
-            _dataProtector = dataProtectionProvider.CreateProtector("MediaProcessingTokenService");
             _memoryCache = memoryCache;
+            _siteService = siteService;
             foreach (IImageWebProcessor processor in processors)
             {
                 foreach (string command in processor.Commands)
                 {
                     _knownCommands.Add(command);
                 }
             }
+
+            var siteSettings = _siteService.GetSiteSettingsAsync().GetAwaiter().GetResult();
+            var mediaTokenSettings = siteSettings.As<MediaTokenSettings>();
+            _hashKey = mediaTokenSettings.HashKey;
```

Comment:
@jtkech I opted here, as it is a singleton, to retrieve the value in the constructor, rather than making every method async, and checking in the methods, whether the value was null, and retrieving it if null.

(some methods that use this would be breaking to make async as well)

Reply:
@deanmarcussen 

Yes `ISiteService` is an exception (see below) as it is a singleton using a scoped `ISession`, so yes you can inject it in a singleton, but at least you should not call its methods outside a scope which may happen in a singleton constructor. 

It works as long as `IMediaTokenService` is injected only by scoped services, but i think it's not good to do that.

Also `SiteSettings` is a shared document but here `mediaTokenSettings` may not be in sync with other running instances. E.g. if a migration completes later, or if you re-generate the key on one instance, currently it is done by disabling / re-enabling the feature so the tenant is rebuilt, but here you would need the distributed tenants feature.

So maybe better to get the key from the cached settings on demand, and maybe generate it if it doesn't exist yet.

Also not sure it's good to use a migration step, maybe a shell activated handler to generate the hash key if it doesn't exist yet. OR as said above just generate the key on demand. As a side note, there are remaining things as migrations that would need to be atomic / exclusive, see #7941, i'm working on, e.g. on WF Singleton.

Then okay for the `IFeatureEventHandler` if the only way to re-generate a key is to disable / re-enable the feature, but why not a custom site settings driver / service to do that, e.g. through a `Generate a new key` button, but still generate it automatically on demand if it doesn't exist yet.

---

The `ISiteService` singleton is an exception as it uses a scoped `ISession`, it is injected by the `HomeRouteTransformer` singleton, hmm just tried to register this tansformer as a scoped service, it works. But it is also injected by an `IConfigureOptions<>` so somewhere by an `IOptions<>` singleton, here we would need to create a scope on the fly, or just retrieve the current one as we do in `ISiteService`.

Maybe worth to try to make `ISiteService` a scoped service to avoid bad usage, i will try it.

Reply:
Ok, so sounds like it's better to get the site settings on demand once, and store the value locally in the singleton, but retrieve outside of the constructor.

Could do shell activating, rather than a migration, but the migration was only a one hit **for existing sites**.

I preferred the `IFeatureEventHandler` because I didn't want to expose it through a settings driver (figuring then people will expect a recipe step, and deployment step for it, when it should really be randomly generated per server - and rarely changed)

Reply:
> Could do shell activating, rather than a migration, but the migration was only a one hit for existing sites.

Yes i know, maybe because i was thinking about making migrations atomic (related to #7941) and didn't want to add complexity here. Note: We already use a local lock for activating a shell (where auto migrations happen)

I was also thinking about another instance generating the key, so anyway we would need to first check if the key already exists before generating a new one, so i thought about the handler to have a better chance to be in sync.

> I preferred the IFeatureEventHandler because I didn't want to expose it through a settings driver

Okay

> when it should really be randomly generated per server - and rarely changed

**Do you mean a different key per server even they all serve the same web app?** If so, we would need to store the key per server, so not through a shared document as site settings. Otherwise, if the key need to be in sync across nodes for a given web app, maybe better to check a shared identifier as we do through shared documents.

Can't right now but i will focus on it this night, but need to know if the key needs to be in sync

Reply:
Thanks @jtkech 

Same key per server. 

Different key for development etc (so it is not important to be able to set the key through a recipe)

It would be a rare requirement to want to change it. (changing it would be a result of a security concern, as it would cause every image to refresh in the browsers cache. Just the browsers cache, not the is-cache. So you would only change it as a result of an attempted attack, that targeted this particular area, i.e. disk filling through images)

So I think site settings is ok for this.

It's random rather than something we set in code, so that it is a per site key (and per tenant).

Now in terms of being in sync between servers.

I assume that activating the feature across a multi-node site would probably trigger the reset off the key twice, but through the shared document sync, only the first one would win? As you have already avoided races there.

Reply:
@deanmarcussen okay cool i understand

> I assume that activating the feature across a multi-node site would probably trigger the reset off the key twice

Yes but only if you enable the Distributed Tenants feature which needs an `IDistributedCache` implementation as `RedisCache`, but that is a separate feature.

> but through the shared document sync, only the first one would win?

Maybe yes if in a given usage we save a new value only if there is no existing one, but it may still happen concurrently. And no if 2 instances want to explicitly save a new value, the last win. That's why we use identifiers (one as a cache entry and one embedded in the document itself) to keep in sync the 2 instances, but each instane need to get data on demand from the multi level cache to keep up to date values.

That's the main goal, keep in sync instances between them and the data store. If we would want that the first win, we would need to use a distributed lock to do an exclusive work.

> As you have already avoided races there.

That's okay here, normally it is thread safe, but as said atomicity / exclusive jobs and so on (not all services need them) are other concerns i'm working on